### ISSUE TITLE
Multiplication of non-commutative and commutative symbols

### ISSUE THREAD

**Issue by klim-** [sympy/sympy#18367]
I am working with both commutative and non-commutative Symbols. I am kind of surprised that the following problem has not been addressed before (at least I have not found it).
```
>>> import sympy as sp
>>> sp.__version__
'1.5.1'
>>> a = sp.Symbol("a")
>>> a.is_commutative
True
>>> b = sp.Symbol("b", commutative=False)
>>> b.is_commutative
False
>>> c = b*a
>>> c
a*b
>>> c.is_commutative
False
>>> d = b*c
>>> d
a*b**2
```
Unfortunately this problem results in very ugly work arounds (casting every symbol to a non-commutative symbol, multiply, and cast back, etc.) that I do not find satisfying.

**Comment by oscarbenjamin** [comment]
Can you clarify the problem?

As far as I can tell you are asking why `b*b*a == a*b**2` when `a` is commutative and `b` is not. In SymPy if `a` is commutative then it commutes with non-commutative objects. Imagine that `a` is a scalar and `b` a matrix: is the transformation not valid in that case?

**Comment by klim-** [comment]
What you are saying is exactly what I was observing. But from a mathematical point of view this does not really make sense. In the general case (i.e. the free group of a set of symbols) `a*b` is not equal `b*a`, and also words like `a*b*a` are not equal to `a**2*b`, etc. Let's say you specify certain symbols to commute with each other, e.g. `a*c=c*a`, then this relation can be substituted in any word containing `c**i*a**j` or `a**i*c**j` (for some `i`,`j`) without changing the element. But since `b` does not commute, `a*b` is still not equal to `b*a`. Non-commutativity is the general case, not the other way around!
And I have to correct you regarding the example: scalars are elements of a field which do not have to commute with the elements of the matrix (think matrices over rings)! Thus scalars and matrices commute if and only if the scalars commute with the elements of the matrix. So your example is a special case and should not be the reason for the current implementation.

**Comment by oscarbenjamin** [comment]
It comes down to the definition. It is unfortunate that in SymPy (currently) commutative is a property of a single object when it should really only be defined between a pair of objects. That is the way it is though so your complaint here comes down to this:
* You would like it to be that if a is commutative and b is not then a does not commute with b and `a*b!=b*a`.
* The current state in SymPy is that if a is commutative and b is not then a and b commute so `a*b==b*a`.

**Comment by klim-** [comment]
In fact, commutativity is defined between (all possible) pairs of symbols already, namely all symbols that are commutative! Non-commutativity is the general case, and this is how sympy should handle the multiplication as well. Your reasoning does not convince me.

**Comment by oscarbenjamin** [comment]
> Your reasoning does not convince me.

I am not trying to convince you. I am just explaining the way that the commutative attribute is currently defined in SymPy and what it is that you are proposing should be changed.

It would be fiddly but not impossible to change this. Clearly though the change would not be backwards compatible for anyone using non-commutative symbols.

It sounds to me like you are in a situation where you are really using a non-commutative algebra of some kind but you want certain elements of the algebra to commute with each other (presumably for cancellation etc). The cleanest way to implement that would be that all of your symbols are declared non-commutative but you make a post-processor function that understands which symbols commute with each other and canonicalises `Mul` and `Pow` accordingly. This is how the `qapply` function in quantum works.

**Comment by klim-** [comment]
Yes, this is my current workaround. But it really is the wrong approach. In mathematics, if multiplication of two objects is allowed, then it should be well defined, and therefore there is no place where you can freely choose which property is stronger.

Sure, the change would not be backwards compatible, but as far as I can see it would not break much either: The current workaround is converting commutative symbols to non-commutative which would not be affected by the changes. On the other hand, this would clean up a lot of unnecessary code, since these routines would be obsolete.

> That is the way it is though

This is a lazy excuse for not recognizing a problem. I believe it is fiddly work though.

**Comment by oscarbenjamin** [comment]
> This is a lazy excuse for not recognizing a problem (I don't mean to be rude!).

Well don't be rude then. This isn't about laziness, it is a question of definition. You want one definition that happens to fit your use case but sympy uses another definition.

The meaning of the commutative attribute in sympy is that if an object has commutative=True then it commutes with *everything*. If an object has commutative=False then it might commute with some things but does not necessarily commute with everything.

That distinction might not be what you want for what you are doing but that does not make it incorrect: there is no mathematical definition of commutativity as a property of an individual object.

It isn't possible for `Mul.flatten` to understand all of the possible commutativity relationships between different types of objects so commutative has been defined so that it can simply separate the commutative and non-commutative parts. If that flattening does not do what you want then you can create a new `Mul` with its own `flatten` (like `MatMul` for matrices).

**Comment by klim-** [comment]
Sorry for being rude, then. I appreciate your time and work.

I understand what is going on and I have an unpleasant workaround. But I question the current sympy interpretation, which I find should be based on the mathematical foundation of the free group of symbols, and not some arbitrary interpretation.

> If an object has commutative=False then it might commute with some things

This is the contradiction I am referring to.

To be a bit more constructive: Do you know where changes would have to be made? Do you have an idea what exactly would break if this change would be implemented?

**Comment by oscarbenjamin** [comment]
> > If an object has commutative=False then it might commute with some things
> 
> This is the contradiction I am referring to.

It's not a contradiction. Commutativity is an *all* property for a set `S` and binary operator `+` that
```
(S, +) is commutative := for all a, b in S: a+b == b+a
```
In sympy this is taken to be true for `+` where the set is `E` the set of all `Expr` instances:
https://github.com/sympy/sympy/blob/f5413ff89530b14ffc9b7359717f8c7ad03d2c6c/sympy/core/add.py#L273-L274

In sympy the commutative attribute is a property of an object and refers only to multiplication so instead we have
```
a in E, a is commutative := for all b in E: a*b == b*a
```
Then non-commutative is the negation of that
```
a in E, a is non-commutative := exists b in E: a*b != b*a
```
It can still be the case that a is non-commutative but there exists some b such that `a*b == b*a` (e.g. if b is commutative then that holds by the definition of commutative above).


> Do you know where changes would have to be made?

The code that handles this is in `Mul.flatten`:
https://github.com/sympy/sympy/blob/f5413ff89530b14ffc9b7359717f8c7ad03d2c6c/sympy/core/mul.py#L112-L114

Note that separating commutatives and non-commutatives is even built in to the signature of the flatten function.

The difficulty with `Add.flatten` and `Mul.flatten` is the fact that so many different Expr subclasses can appear here. Personally I think it would be better if there was no special case handling for non-commutatives and Order here. It would be better if this class was only used for expressions representing complex numbers and infinities. Anything else should have another class/routine e.g. `NCMul` for handling non-commutatives.

**Comment by oscarbenjamin** [comment]
> Do you have an idea what exactly would break if this change would be implemented?

No but this is such a core routine that you will see a lot of test-failures if you change anything. I recommend trying something out and running the test suite. That will show you what is depending on this behaviour.

**Comment by klim-** [comment]
> > Do you have an idea what exactly would break if this change would be implemented?
> 
> No but this is such a core routine that you will see a lot of test-failures if you change anything. I recommend trying something out and running the test suite.

This is a helpful, thanks!

And the other way around: What would break when getting rid of the non-commutative option as it is now, altogether, and then re-implement this from the ground up (e.g. `b = sp.NcSymbol("b")`)? I could imagine this breaks a lot less, and a transition from the current state would also be easier (well, I guess I could check that using the tests as well...).

**Comment by sylee957** [comment]
I think that this the limitation in the current architecture of assumptions, because `commutative` is defined in some universe, and it is not possible to define something like `a.commutes_with(b)` as assumptions.

I guess that some internal simplification algorithms should check commutativity binarily, and it can be more useful than downright considering identity matrix is not commutative.

**Comment by oscarbenjamin** [comment]
The identity matrix is not really commutative because e.g.:
```julia
In [14]: A = eye(2)                                                                                                                            

In [15]: B = zeros(2, 3)                                                                                                                       

In [16]: A*B                                                                                                                                   
Out[16]: 
⎡0  0  0⎤
⎢       ⎥
⎣0  0  0⎦

In [17]: B*A                                                                                                                                   
---------------------------------------------------------------------------
ShapeError
```
This is all handled by `MatMul` though which can understand not just whether a matrix is the identity matrix but also the shapes etc. That is the right way to handle this: dedicated code that understands the kind of objects that it is working with.

**Comment by mcditoos** [comment]
In order not to open another issuer is it possible currently to have b= sp.Symbol("b", commutative=False) commute with  a=sp.Symbol("a", commutative=False), besides the workaround of using .subs(a*b,b*a)?
